fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/TableCollector.java

```diff
@@ -17,53 +17,97 @@
+ * Materialized view is extended or not can be controlled by the field expand
  */
-public class TableCollector extends DefaultPlanVisitor<Void, TableCollectorContext> {
+public class TableCollector extends DefaultPlanVisitor<Plan, TableCollectorContext> {
 
     public static final TableCollector INSTANCE = new TableCollector();
+    private static final Logger LOG = LogManager.getLogger(TableCollector.class);
 
     @Override
-    public Void visit(Plan plan, TableCollectorContext context) {
-        if (plan instanceof CatalogRelation) {
-            TableIf table = ((CatalogRelation) plan).getTable();
-            if (context.getTargetTableTypes().isEmpty() || context.getTargetTableTypes().contains(table.getType())) {
-                context.getCollectedTables().add(table);
-            }
+    public Plan visitLogicalCatalogRelation(LogicalCatalogRelation catalogRelation, TableCollectorContext context) {
+        TableIf table = catalogRelation.getTable();
+        if (context.getTargetTableTypes().isEmpty() || context.getTargetTableTypes().contains(table.getType())) {
+            context.getCollectedTables().add(table);
+        }
+        if (table instanceof MTMV) {
+            expandMvAndCollect((MTMV) table, context);
+        }
+        return catalogRelation;
+    }
+
+    @Override
+    public Plan visitPhysicalCatalogRelation(PhysicalCatalogRelation catalogRelation, TableCollectorContext context) {
+        TableIf table = catalogRelation.getTable();
+        if (context.getTargetTableTypes().isEmpty() || context.getTargetTableTypes().contains(table.getType())) {
+            context.getCollectedTables().add(table);
+        }
+        if (table instanceof MTMV) {
+            expandMvAndCollect((MTMV) table, context);
+        }
+        return catalogRelation;
+    }
+
+    private void expandMvAndCollect(MTMV mtmv, TableCollectorContext context) {
+        if (!context.isExpand()) {
+            return;
+        }
+        try {
+            MTMVCache expandedMv = MTMVCache.from(mtmv, MTMVPlanUtil.createMTMVContext(mtmv));
+            expandedMv.getLogicalPlan().accept(this, context);
+        } catch (AnalysisException e) {
+            LOG.error(String.format(
+                    "table collector expand fail, mtmv name is %s, targetTableTypes is %s",
+                    mtmv.getName(), context.targetTableTypes), e);
         }
-        return super.visit(plan, context);
     }
 
     /**
      * The context for table collecting, it contains the target collect table types
      * and the result of collect.
      */
     public static final class TableCollectorContext {
-        private final List<TableIf> collectedTables = new ArrayList<>();
+        private final Set<TableIf> collectedTables = new HashSet<>();
         private final Set<TableType> targetTableTypes;
+        // if expand the mv or not
+        private final boolean expand;
```

Comment:
always true now?

Reply:
this method provides an ability which can set expland true or false. in the scene， it's should always be true